### PR TITLE
#610 /start with arguments

### PR DESCRIPTION
**Pull request by komaz**
Proposed fix for issue #610 – `TkStart` now checks if 'post' and 'invite' parameters are specified and performs required actions. `TkStartTest.postsMessageAndInvitesFriends` added.

### CODE REVIEW

**Comment by karato** [comment]
@komaz Thanks, let me find someone who can review this pull request

**Comment by karato** [comment]
@caarlos0 review it

File: netbout-web/src/main/java/com/netbout/rest/TkStart.java [hunk 1]
```diff
@@ -59,11 +65,34 @@ public TkStart(final Base bse) {
...
+            String.format("new bout #%d started", number)
+        );
+        final Href href = new RqHref.Base(req).href();
+        final Iterator<String> invite = href.param("invite").iterator();
+        while (invite.hasNext()) {
+            final String friend = invite.next();
+            try {
+                bout.friends().invite(friend);
+            } catch (final Friends.UnknownAliasException ex) {
+                throw new RsFailure(ex);
```
Comment (caarlos0) [inline netbout-web/src/main/java/com/netbout/rest/TkStart.java]: at this point the bout was already created, right? Shouldn't  we do something about it?

File: netbout-web/src/main/java/com/netbout/rest/TkStart.java [hunk 2]
```diff
@@ -59,11 +65,34 @@ public TkStart(final Base bse) {
...
+            }
+            msg.append(String.format(", \"%s\" invited", friend));
+        }
+        final Iterator<String> post = href.param("post").iterator();
+        if (post.hasNext()) {
+            try {
+                bout.messages().post(post.next());
+            } catch (final Messages.BrokenPostException ex) {
+                throw new RsFailure(ex);
```
Comment (caarlos0) [inline netbout-web/src/main/java/com/netbout/rest/TkStart.java]: @komaz same as pointed in previous comment

File: netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        final Message message = bout.messages().iterate().iterator().next();
+        MatcherAssert.assertThat(message.author(), Matchers.equalTo(jeff));
+        MatcherAssert.assertThat(message.text(), Matchers.equalTo(post));
+        MatcherAssert.assertThat(
+            bout.friends().iterate().iterator().next().alias(),
+            Matchers.equalTo(frank)
+        );
+    }
+
+}
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java]: @komaz I think you are missing tests for the error scenarios, like when a friend is not found...

**Comment by caarlos0** [comment]
@komaz just a couple of comments above

**Comment by komaz** [comment]
@caarlos0 Thanks for the review. Regarding bout removal – I don't see any way to remove a bout, so I added dropping all users from it, do you think it is ok?
Regarding negative tests – I found an issue #628, and it seems to me that adding tests for negative scenarios here is way over the budget. What is the correct way to track negative tests as a separate task?

**Comment by caarlos0** [comment]
@komaz looks nice.

About the tests: I believe you should have add the tests first, how can you guarantee your code work without the right coverage...

Anyways, I guess you can add a TODO for that...

**Comment by komaz** [comment]
@caarlos0 thank you, I'll make sure I always start with tests for future tasks. Please take a look at added `@todo`, does it look good to you?

File: netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            bout.friends().iterate().iterator().next().alias(),
+            Matchers.equalTo(frank)
+        );
+    }
+
+    /**
+     * TkStart can remove all friends from a bout when some of invitees
+     * is not found.
+     * @throws Exception If there is some problem inside
+     * @todo #610 negative test is not implemented
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java]: @komaz let's put the TODO in the class javadoc and remove this empty method. Also, TODOs should start with `#ISSUE_NUMBER:TIME_INmin` and have at least 20 words, or pdd will complain.

**Comment by dmzaytsev** [comment]
@caarlos0 @komaz I propose to remove `TkStart.discard()` to avoid untested code.
We can make puzzle in `TkStart` for a "bout rollback" if an exception occurs. 
what do you think?

**Comment by caarlos0** [comment]
@dmzaytsev I would say to leave it there (as you already written it) and just add the TODO...

**Comment by dmzaytsev** [comment]
@caarlos0 if we leave `discard()` there why we need add one more TODO?

File: netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        friend.aliases().add(frank);
+        final String post = "message";
+        final String head = new RsPrint(
+            new TkForward(new TkStart(base)).act(
+                new RqWithTester(
+                    urn,
+                    new RqFake(
+                        "GET",
+                        String.format(
+                            "/start?post=%s&invite=%s",
```
Comment (dmzaytsev) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java]: @komaz please add more then one `invite` param as requested in the original ticket

**Comment by komaz** [comment]
@caarlos0, moved `@todo` to class level, please take a loo
@dmzaytsev, changed test to invite 2 people into bout

**Comment by caarlos0** [comment]
@dmzaytsev the TODO is about testing, not about the feature itself

**Comment by caarlos0** [comment]
@rultor good to merge

**Comment by caarlos0** [comment]
> @caarlos0, moved @todo to class level, please take a loo

@komaz looks good, thanks

**Comment by dmzaytsev** [comment]
@caarlos0 @komaz the Travis build failed...
we can't merge this PR right now

**Comment by komaz** [comment]
@dmzaytsev looks like **master** is broken – https://travis-ci.org/yegor256/netbout/builds/63865407

**Comment by dmzaytsev** [comment]
@komaz let's wait #636

**Comment by dmzaytsev** [comment]
@rultor try to merge

**Comment by dmzaytsev** [comment]
@komaz please fix the puzzle description
> Space expected in line #53 in /home/r/repo/netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java (PDD::Error)

File: netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+import org.takes.facets.forward.TkForward;
+import org.takes.rq.RqFake;
+import org.takes.rs.RsPrint;
+
+/**
+ * Test case for {@link TkStart}.
+ * @author Ivan Inozemtsev (ivan.inozemtsev@gmail.com)
+ * @version $Id$
+ * @todo #610:30min Test case for negative scenario of TkStart
+ * should be added, e.g. when a friend is not found, the error
```
Comment (dmzaytsev) [inline netbout-web/src/test/java/com/netbout/rest/bout/TkStartTest.java]: @komaz one space before `should`
same two lines below

**Comment by komaz** [comment]
@dmzaytsev thanks, it would take me a while to understand where to add missing spaces

**Comment by dmzaytsev** [comment]
@rultor merge please

**Comment by karato** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy

**Comment by ypshenychka** [comment]
@karato Quality is good.

**Comment by karato** [comment]
> @karato Quality is good.
> 

@ypshenychka many thanks

**Comment by karato** [comment]
@caarlos0 I added 10 mins to @ypshenychka (for QA review) in transaction 58143590; since quality is good, 10 mins paid to @dmzaytsev (our architect), transaction `58143605`; paid, thanks, added **21 mins** to your account, payment `AP-3941464256754212X`, 105 hours and 24 mins was spent; you're getting extra minutes here (c=6); +21 added to your rating, current score is: [+2146](http://www.netbout.com/b/35095?open=rating)
